drivers: pwm: nrfx: 0/100% duty for PWM120

mstasiaknordic opened the pull request:
Fast PWM120 prevents GPIO from driving pin with low/high state when PWM duty is 0% or 100%. In such case, regular PWM signal generation needs to be used.

nika-nordic commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -234,7 +234,13 @@ static int pwm_nrfx_set_cycles(const struct device *dev, uint32_t channel,
 
+#if NRF_PWM_HAS_IDLEOUT
+			nrfx_pwm_stop(&config->pwm, true);
+			nrfy_pwm_channel_idle_set(config->pwm.p_reg, channel, out_level);
```
why can't you drive P7 pin directly? because of CTRLSEL==PWM ?

nika-nordic commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -234,7 +234,13 @@ static int pwm_nrfx_set_cycles(const struct device *dev, uint32_t channel,
 			}
 
+#if NRF_PWM_HAS_IDLEOUT
+			nrfx_pwm_stop(&config->pwm, true);
```
isn't the PWM already stopped at this point?

mstasiaknordic replied:
Not if previous sequence actually used PWM. Moreover, `nrfy_pwm_disable` is called within irq handler when used with `nrfx_pwm_stop(..., false)`. In this case, using `nrfx_pwm_stop(..., true)` makes sure that PWM is disabled and driver is set to initialized state before setting the value of idleout, which would normally be ignored.

nordic-krch commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -234,7 +234,13 @@ static int pwm_nrfx_set_cycles(const struct device *dev, uint32_t channel,
 			}
 
+#if NRF_PWM_HAS_IDLEOUT
+			nrfx_pwm_stop(&config->pwm, true);
```
What happens if another channel is generating a pulse. Won't this stop that second channel?

mstasiaknordic replied:
Unfortunately yes, the who device needs to be disabled in order to change the value of idleout for any channel

nordic-krch replied:
So i don't think that this is acceptable. When idleout is needed? Can't we use gpio in case when there are active PWM channels  and set idleout for all channels only if all channels have duty cycle 0 or 100?

mstasiaknordic replied:
GPIO is the best solution, however, it does not work on fast instances, where it cannot drive the pin, unless there is a way of granting ownership of the pin to the GPIO during runtime. I agree that such solution is imperfect, especially since it is not the purpose of idleout mechanism, however, it is the only way to get 0/100 duty cycle

nika-nordic replied:
using GPIO registers when dedicated pins are utilized won't work because in such case pins are wired directly to the peripheral through UICR configuration.

Maybe if setting IDLEOUT would break another active PWM channel we need to return an error to the user, if 0 or 100 duty cycle is requested?

nordic-krch replied:
> Maybe if setting IDLEOUT would break another active PWM channel we need to return an error to the user

It would be hard for user to handle that error as from PWM API perspective it should be supported (and was in all other nRF devices).

Can't we just use PWM to use 0 or 100 duty cycle? If there is any other active PWM channel then PWM is running anyway so it will not impact power consumption. Only when no channel is needing PWM we can set idleout on every channel.

mstasiaknordic committed:
drivers: pwm: nrfx: 0/100% duty for PWM120

Fast PWM120 prevents GPIO from driving pin with low/high
state when PWM duty is 0% or 100%. In such case, regular
PWM signal generation needs to be used.

Signed-off-by: Michał Stasiak <michal.stasiak@nordicsemi.no>

mstasiaknordic commented:
Decided not to use IDLEOUT feature, since it would require stopping PWM for a short period of time, even if no signal is generated. which would result in ~5us of low state on all channels. 

In current implementation, PWM120 is used to generate all signals, and if all of them happen to be either 0% o 100% duty, high frequency clock is released.

mstasiaknordic commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -86,6 +86,11 @@ struct pwm_nrfx_data {
 #error "Current implementation supports maximum 8 channels."
 #endif
 
+static bool pwm_is_fast(const struct pwm_nrfx_config *config)
+{
+	return config->clock_freq > (16ul * 1000ul * 1000ul);
+}
+
```
Needed a function to runtime check whether an instance is fast.

bjarki-andreasen commented:
ping @masz-nordic

masz-nordic commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -86,6 +86,11 @@ struct pwm_nrfx_data {
 
+static bool pwm_is_fast(const struct pwm_nrfx_config *config)
+{
+	return config->clock_freq > (16ul * 1000ul * 1000ul);
```
Nit:
```suggestion
	return config->clock_freq > MHZ(16);
```

masz-nordic commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -86,6 +86,11 @@ struct pwm_nrfx_data {
 #error "Current implementation supports maximum 8 channels."
 #endif
 
+static bool pwm_is_fast(const struct pwm_nrfx_config *config)
```
```
#if NRFX_FOREACH_PRESENT(PWM, PWM_NRFX_IS_FAST, (||), (0)))
...
#else
static bool pwm_is_fast(const struct pwm_nrfx_config *config)
{
return false;
}
```

masz-nordic replied:
This is just an optimization. I would expect compiler be able to optimize out the `if` in line 271 and 291 with that change.

mstasiaknordic replied:
Okay, then I guess its worth adding

masz-nordic commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -259,8 +264,12 @@ static int pwm_nrfx_set_cycles(const struct device *dev, uint32_t channel,
-
-			nrf_gpio_pin_write(psel, out_level);
+			if (pwm_is_fast(config)) {
+				nrfx_pwm_simple_playback(&config->pwm, &config->seq, 1,
```
Please add comment on why this is different for fast PWM.

nika-nordic commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -259,8 +264,12 @@ static int pwm_nrfx_set_cycles(const struct device *dev, uint32_t channel,
 			}
-
-			nrf_gpio_pin_write(psel, out_level);
+			if (pwm_is_fast(config)) {
```
I would suggest to put a comment here why `fast` instance cannot utilize GPIO write:
```
Fast PWM instances output is connected directly to GPIO pads hence cannot be controlled by GPIO registers directly. Use 100% duty cycle playback instead.
```

nika-nordic commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -259,8 +264,12 @@ static int pwm_nrfx_set_cycles(const struct device *dev, uint32_t channel,
-
-			nrf_gpio_pin_write(psel, out_level);
+			if (pwm_is_fast(config)) {
+				nrfx_pwm_simple_playback(&config->pwm, &config->seq, 1,
```
shouldn't we modify `&config->seq` for given channel to be sure it will have 100% duty cycle before starting the playback?

mstasiaknordic replied:
It is done no matter the duty here :

`seq_values_ptr_get(dev)[channel] = PWM_NRFX_CH_VALUE(compare_value, inverted);`

nika-nordic commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -275,6 +284,22 @@ static int pwm_nrfx_set_cycles(const struct device *dev, uint32_t channel,
+		if (pwm_is_fast(config)) {
+#if PWM_NRFX_USE_CLOCK_CONTROL
+			if (data->clock_requested) {
+				int ret = nrf_clock_control_release(config->clk_dev,
```
in the diff above playback is started for `!needs_pwm` conditional, so I guess you cannot release clock here, or it would disrupt the 100% duty cycle playback

mstasiaknordic replied:
From my understanding, if `needs_pwm==0` then all outputs are either 0% or 100%, none of them are actual PWM sequences. In that case, compare value is either 0 or equal to the COUNTERTOP, so frequency of the counter does not matter and thus the clock can be released.

nika-nordic replied:
so if we release the clock, the clock itself will still be available but with unknown frequency, correct ?

nordic-krch replied:
when `data->clock_requested == true` doesn't it mean that PWM instance is fast and `pwm_is_fast` check is redundant?

nordic-krch commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -259,8 +271,16 @@ static int pwm_nrfx_set_cycles(const struct device *dev, uint32_t channel,
 			}
-
-			nrf_gpio_pin_write(psel, out_level);
+			/* Output of fast PWM instance is directly connected to GPIO pads,
```
what about having that code only when needed and utilizing a fact that only fast instance has `clk_dev`?
```
#if PWM_NRFX_USE_CLOCK_CONTROL
if (config->clk_dev) {
  nrfx_pwm_simple_playback(...);
} else
#endif 
{
  nrf_gpio_pin_write(psel, out_level);
}
```

mstasiaknordic replied:
I agree if `PWM_NRFX_USE_CLOCK_CONTROL` symbol was to be replaced with one that indicates presence of fast instance, probably equal to `NRFX_FOREACH_PRESENT(PWM, PWM_NRFX_IS_FAST, (||), (0))`.

nordic-krch commented on drivers/pwm/pwm_nrfx.c:
```diff
@@ -86,6 +86,18 @@ struct pwm_nrfx_data {
+#if NRFX_FOREACH_PRESENT(PWM, PWM_NRFX_IS_FAST, (||), (0))
+static bool pwm_is_fast(const struct pwm_nrfx_config *config)
+{
+	return config->clock_freq > MHZ(16);
```
`config->clk_dev != NULL` would be less code.

mstasiaknordic replied:
`clk_dev` field is present only if fast instance is enabled and `CONFIG_CLOCK_CONTROL_NRF2_GLOBAL_HSFLL=y`. However, issues with driving pins on fast instances occur with clock control disabled as well.

nordic-krch replied:
but isn't use of hsfll mandatory when fast instance is used?

mstasiaknordic replied:
You are right, I was wrongly convinced that `CONFIG_CLOCK_CONTROL_NRF2_GLOBAL_HSFLL` was n by default. I agree then

mstasiaknordic replied:
Apparently not, as `CONFIG_CLOCK_CONTROL` needs to explicitly enabled to use global hsfll request/release. So in my opinion, the case of using fast instances without clock control needs to be covered
